mozilla/bedrock
# 10867: Support syncing multi-locale Contentful entries to Bedrock's DB

## Description

This changeset extends the current pattern of storing Contentful "pages" (`Entry` records and their children) as Django `ContentfulEntry` records, setting the appropriate `locale` field. 



## Issue / Bugzilla link

Contributes to #10867

## Testing - assuming local Docker usage

* Set `CONTENTFUL_ENVIRONMENT=sandbox` in your  `.env`. 
* Restart your `app` container
* `make shell` 
* `./manage.py update_contentful --force`

The sandbox env is set up to have just three locales enabled (`en-US`, `de-DE`, `fr-FR`) and two of the VRC pages also have mock German translations (en-US text with "GERMAN" written above, and one of these pages also has a French translation. All should sync down to your local database, which you can check in a Django shell:

```
$ ./manage.py update_contentful --force
Running forced update from Contentful data
Pulling from Contentful
Page connectHomepage:58YIvwDmzSDjREDACTED deemed not syncable for de-DE    # Homepages only synced with no-op en-US locale
Page connectHomepage:4k3CxqZGjxXOREDACTED deemed not syncable for de-DE
Page connectHomepage:58YIvwDmzSDjREDACTED deemed not syncable for fr-FR
Page connectHomepage:4k3CxqZGjxXOREDACTED deemed not syncable for fr-FR
Updating existing ContentfulEntry for connectHomepage:en-US:58YIvwDmzSDjREDACTED  # ie, synced here.
Updating existing ContentfulEntry for connectHomepage:en-US:4k3CxqZGjxXOREDACTED
Updating existing ContentfulEntry for pagePageResourceCenter:en-US:1dQAhsz5XnrX6IREDACTED
[...]
Problem with page:3bKxRzCv6kHjCAREDACTED -> <class 'Exception'>: No 'Entry' detected for page   # bad data in sandbox
Updating existing ContentfulEntry for pagePageResourceCenter:en-US:IORNDWEoanCkQREDACTED
Updating existing ContentfulEntry for pagePageResourceCenter:en-US:5wIcW1esdWIK6REDACTED
Locale mismatch on pageHome:1XWEqmSvQQF6eREDACTED -> en-US vs Untitled page - 📄 Page: Home   # bad data in sandbox
Creating new ContentfulEntry for pagePageResourceCenter:en-US:2LEDujIfxanZKMREDACTED
Locale mismatch on pageHome:60YAs15SYSKKnjREDACTED -> en-US vs Untitled page - 📄 Page: Home   # bad data in sandbox
Locale mismatch on pageHome:1HercJWjhIEm4SREDACTED -> en-US vs Untitled page - 📄 Page: Home   # bad data in sandbox
Creating new ContentfulEntry for pagePageResourceCenter:de-DE:1dQAhsz5XnrX6IREDACTED
Creating new ContentfulEntry for pagePageResourceCenter:de-DE:4Gobq3NdbOCaaaREDACTED
No content for 1JxYJUTGphwTNRREDACTED for de-DE
No content for 3gikThlF98f2heREDACTED for de-DE
[...]
Creating new ContentfulEntry for pagePageResourceCenter:fr-FR:1dQAhsz5XnrX6IREDACTED
No content for 4Gobq3NdbOCaaaREDACTED for fr-FR
[...]
No content for 1HercJWjhIEm4SREDACTED for fr-FR
Entries to be deleted: <QuerySet []>
Deleted by _detect_and_delete_absent_entries: (0, {'contentful.ContentfulEntry': 0})
Done. Added: 4. Updated: 8. Deleted: 0. Errors: 4
```

##### Things to experiment with

Fire up Django shell and try out these fun things! They're most easily checked by running ORM queries for `bedrock.contentful.models.ContentfulEntry`

* Archive an entire multi-locale page and check what happens when you force-sync locally. Do the relevant pages get deleted? Do the remaining pages persist?
* Unarchive the archived page and re-sync - all the pages will be back
* Edit a multi-locale Compose page in Contentful and remove the SEO and main content entries/children for a particular locale. Force resync - the speciifc zapped locale should be deleted but the pages other locales should remain
* Restore/re-link the missing data from the previous step (bear in mind how we model SEO objects may change once we get data back from translation, so go with the triple deck of SEO objects for now). Re-sync. Nice to have the data back, yep?
